sleap/gui/commands.py

```diff
@@ -3750,6 +3751,67 @@
 
         return views
 
+    @staticmethod
+    def get_permutations_of_instances(
+        selected_instance: Instance,
+        session: RecordingSession,
+        frame_idx: int,
+        cams_to_include: Optional[List[Camcorder]] = None,
+    ) -> Dict[int, Dict[Camcorder, List[Instance]]]:
+        """Get all (single-instance) possible combinations of instances across views.
+
+        Args:
+            selected_instance: The `Instance` to add to  permutations of instances in
+                views other than that of the `selected_instance`.
+            session: The `RecordingSession` containing the `Camcorder`s.
+            frame_idx: Frame index to get instances from (0-indexed).
+            cams_to_include: List of `Camcorder`s to include. Default is all.
+            require_multiple_views: If True, then raise and error if one or less views
+                or instances are found.
+
+        Raises:
+            ValueError if one or less views or instances are found.
+
+        Returns:
+            Dict with frame identifier keys (not the frame index) and values of another
+            inner dict with `Camcorder` keys and `List[Instance]` values. Each
+            `List[Instance]` is of length 1.
+        """
+
+        cam_selected = session.get_camera(selected_instance.video)
+        cam_selected = cast(Camcorder, cam_selected)  # Could be None if not in session
+
+        # Get all instances accross views at this frame index, then remove selected
+        instances_excluding_selected: Dict[
+            Camcorder, List[Instance]
+        ] = TriangulateSession.get_instances_across_views(
+            session=session,
+            frame_idx=frame_idx,
+            cams_to_include=cams_to_include,
+            track=-1,  # Get all instances regardless of track.
+            require_multiple_views=True,
+        )
+        instances_excluding_selected.pop(cam_selected)
+
+        # Permute instances from other views into all possible combos
+        # Ordering of dict_values is preserved in Python 3.7+
+        permutated_instances: Iterator[Tuple] = product(
+            *instances_excluding_selected.values()
+        )
+
+        # Reorganize permutaions by cam and add selected instance to each permutation
+        instances: Dict[int, Dict[Camcorder, List[Instance]]] = {}
+        for frame_id, perm in enumerate(permutated_instances):
+            instances[frame_id] = {cam_selected: [selected_instance]}
+            instances[frame_id].update(
+                {
+                    cam: [inst]
+                    for cam, inst in zip(instances_excluding_selected.keys(), perm)
+                }
+            )
+
+        return instances  # Expect <num instances in other views>! frames
```

Comment:
What data structure would you recommend?

Reply:
So the reason I need to reorganize is because itertools.product yields a list of tuples where each tuple in the list is a possible frame and each element in the tuple is the instance at camera/view. I am using a dictionary of dictionaries as soon as I can. Do you think it would be faster to not use itertools.product and try to create permutations on my own?